upup/pkg/fi/cloudup/hetznertasks/firewall.go

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hetznertasks
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strconv"
+
+	"github.com/hetznercloud/hcloud-go/hcloud"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kops/upup/pkg/fi/cloudup/hetzner"
+)
+
+// FirewallRule represents a Firewall's rules.
+type FirewallRule struct {
+	Direction string
+	SourceIPs []string
+	Protocol  string
+	Port      *string
+}
+
+// +kops:fitask
+type Firewall struct {
+	Name      *string
+	Lifecycle fi.Lifecycle
+
+	ID       int
+	Selector string
+	Rules    []*FirewallRule
+
+	Labels map[string]string
+}
+
+var _ fi.CompareWithID = &Firewall{}
+
+var _ fi.HasDependencies = &FirewallRule{}
+
+func (e *FirewallRule) GetDependencies(tasks map[string]fi.Task) []fi.Task {
+	return nil
+}
+
+func (v *Firewall) CompareWithID() *string {
+	return fi.String(strconv.Itoa(v.ID))
+}
+
+func (v *Firewall) Find(c *fi.Context) (*Firewall, error) {
+	cloud := c.Cloud.(hetzner.HetznerCloud)
+	client := cloud.FirewallClient()
+
+	firewalls, err := client.All(context.TODO())
+	if err != nil {
+		return nil, err
+	}
+
+	for _, firewall := range firewalls {
+		if firewall.Name == fi.StringValue(v.Name) {
+			matches := &Firewall{
+				Lifecycle: v.Lifecycle,
+				Name:      fi.String(firewall.Name),
+				ID:        firewall.ID,
+				Selector:  firewall.AppliedTo[0].LabelSelector.Selector,
+				Labels:    firewall.Labels,
+			}
+			for _, rule := range firewall.Rules {
```

Comment:
Ah, I see .. the API assumes a strict ordering.  This approach makes sense then, we might need to improve the tasks logic to cope with complex structures!